File: juweirat-api/src/Juweirat.Application/Common/Pagination/PagedResult.cs
namespace Juweirat.Application.Common.Pagination;

public class PagedResult<T>
{
    public IReadOnlyList<T> Items { get; init; } = [];
    public int PageNumber { get; init; }
    public int PageSize { get; init; }
    public int TotalCount { get; init; }
    public int TotalPages { get; init; }
    public bool HasPreviousPage => PageNumber > 1;
    public bool HasNextPage => PageNumber < TotalPages;

    public PagedResult() { }

    public PagedResult(IReadOnlyList<T> items, int count, int pageNumber, int pageSize)
    {
        TotalCount = count;
        PageNumber = pageNumber < 1 ? 1 : pageNumber;
        PageSize = pageSize < 1 ? 1 : pageSize;
        TotalPages = PageSize > 0 ? (int)Math.Ceiling(count / (double)PageSize) : 0;
        Items = items;
    }

    public PagedResult<TDto> Map<TDto>(Func<T, TDto> mapper)
    {
        var mappedItems = Items.Select(mapper).ToList();
        return new PagedResult<TDto>(mappedItems, TotalCount, PageNumber, PageSize);
    }

    public static PagedResult<T> Empty(int pageNumber = 1, int pageSize = 20)
        => new([], 0, pageNumber, pageSize);
}

File: juweiratapp/lib/core/models/paged_result.dart
int _toInt(dynamic value, [int defaultValue = 0]) {
  if (value == null) return defaultValue;
  if (value is int) return value;
  if (value is num) return value.toInt();
  if (value is String) {
    final n = num.tryParse(value);
    if (n != null) return n.toInt();
  }
  return defaultValue;
}

class PagedResult<T> {
  final List<T> items;
  final int pageNumber;
  final int pageSize;
  final int totalCount;
  final int totalPages;
  final bool hasPreviousPage;
  final bool hasNextPage;

  const PagedResult({
    required this.items,
    required this.pageNumber,
    required this.pageSize,
    required this.totalCount,
    required this.totalPages,
    required this.hasPreviousPage,
    required this.hasNextPage,
  });

  factory PagedResult.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic json) fromJsonT,
  ) {
    final rawList = json['items'] as List<dynamic>? ?? [];
    return PagedResult<T>(
      items: rawList.map((e) => fromJsonT(e)).toList(),
      pageNumber: _toInt(json['pageNumber'], 1),
      pageSize: _toInt(json['pageSize'], 10),
      totalCount: _toInt(json['totalCount'], 0),
      totalPages: _toInt(json['totalPages'], 0),
      hasPreviousPage: json['hasPreviousPage'] as bool? ?? false,
      hasNextPage: json['hasNextPage'] as bool? ?? false,
    );
  }
}
